src/StoreApi/Routes/Checkout.php

```diff
@@ -139,71 +139,86 @@ protected function get_route_update_response( \WP_REST_Request $request ) {
+		$order = $this->create_or_update_draft_order();
 
-		if ( ! $order_object instanceof \WC_Order ) {
-			throw new RouteException(
-				'woocommerce_rest_checkout_invalid_order',
-				__( 'This session has no orders pending payment.', 'woo-gutenberg-products-block' ),
-				500
-			);
-		}
+		// --
+		// 2. Process Request
 
-		// Ensure order still matches cart.
-		$order_controller->update_order_from_cart( $order_object );
+		// If any form fields were posted, update the order (e.g. order note, payment method).
+		$this->update_order_from_request( $order, $request );
 
-		// Create a new user account as necessary.
-		// Note - CreateAccount class includes feature gating logic (i.e. this
-		// may not create an account depending on build).
-		if ( defined( 'WC_VERSION' ) && version_compare( WC_VERSION, '4.7', '>=' ) ) {
-			// Checkout signup is feature gated to WooCommerce 4.7 and newer;
-			// Because it requires updated my-account/lost-password screen in 4.7+
-			// for setting initial password.
-			try {
-				$create_account = Package::container()->get( CreateAccount::class );
-				$create_account->from_order_request( $request );
-				$order_object->set_customer_id( get_current_user_id() );
-			} catch ( Exception $error ) {
-				$this->handle_error( $error );
-			}
+		// Validate payment method before proceeding.
+		if ( $order->needs_payment() ) {
+			// Will throw and return 400 if payment method missing or invalid.
+			$this->get_request_payment_method_id( $request );
 		}
-		// If any form fields were posted, update the order.
-		$this->update_order_from_request( $order_object, $request );
 
-		// Check order is still valid.
-		$order_controller->validate_order_before_payment( $order_object );
+		// --
+		// 3. Process Customer
 
-		// Persist customer address data to account.
-		$order_controller->sync_customer_data_with_order( $order_object );
+		// Update order with customer details, and sign up a user account as necessary.
+		$this->process_customer( $order, $request );
+
+		// --
+		// 4. Validate Order
+
+		// Check order is still valid.
+		$order_controller = new OrderController();
+		$order_controller->validate_order_before_payment( $order );
 
 		/*
-		* Fire woocommerce_blocks_checkout_order_processed, should work the same way as woocommerce_checkout_order_processed
-		* But we're opting for a new action because the original ones attaches POST data.
-		* NOTE: this hook is still experimental, and might change or get removed.
-		* @todo: Document and stabilize __experimental_woocommerce_blocks_checkout_order_processed
+		* Action: woocommerce_blocks_checkout_order_processed (experimental).
+		* This hook informs extensions that $order has completed processing and is ready
+		* for payment.
+		*
+		* This is similar to existing core hook woocommerce_checkout_order_processed.
+		* We're using a new action:
+		* - To keep the interface focused (only pass $order, not passing request data).
```

Comment:
I think this is a good initial start. I do wonder however if some extensions might need the request data for any additional information that might have come through in the request. That's unknown at this point though and not something we expose as an option client side for the moment.

Reply:
Yes, this could happen, the original hook passes `$_POST` data.
However, passing $_POST was required because people who render anything in the checkout form and passing `POST`  would provide a way to catch it.
We're not planning on doing that, and since we do everything via REST, any passed back data should via an explicit field and via an explicit mechanism.
Any core data in `$request` that is not present in the order is data we might not want to expose, and exposing `$request` might be too much implementation detail.
So if there's something that we need from `$request` that is not present in order, we can expose it explicitly in the hook.

Reply:
Right, I'm not suggesting we change this signature now. Just thought it was important to note that there may be need for passing along additional information from the request that might not be exposed on the order. It'd be good to think of how we could do that (but not for this PR).